### DISCUSSION TITLE
How to extend ElementView in TS?

### DISCUSSION THREAD

**Discussion by igo** [clientIO/joint#1817] [Q&A] [answered]
I have a custom `Element` and `ElementView` like this:

```ts
class MyElement extends dia.Element {
    myModelMethod() {...}
}
class MyElementView extends dia.ElementView {
    myMethod() {
        this.model.myModelMethod(); // myModelMethod is unknown but works
    }
}
```

However if you use `this.model` in `MyElementView` it is not of type `MyElement`. Is it possible to set type of the model to have all benefits of Typescript?

**Comment by kumilingus** [comment, marked as answer]
You can redefine properties and methods in your custom view if needed.

```ts
class MyElement extends dia.Element {
    test() { return; }
}

class MyElementView extends dia.ElementView {
    model: MyElement;
}

const view = new MyElementView();
view.model.test();
```

Ideally, the ElementView should be a generic class. [This](https://github.com/clientIO/joint/blob/master/types/joint.d.ts#L889) could be perhaps changed to:

```ts
class ElementView<T extends Element = Element> extends CellViewGeneric<T> {
}
```

But more testing would have to be done to make sure it's not a breaking change (See [this](https://stackoverflow.com/questions/73556247/generic-constraints-are-ignored-when-inferring-class-parameters) for more info).

**Reply by igo** [reply]
Thanks. This works:
```ts
model!: MyElement;
```
I totally agree that generics should be used here to enable proper OOP.
